ddl/cluster.go

```diff
@@ -223,17 +265,122 @@ func GetFlashbackKeyRanges(sess sessionctx.Context, startKey kv.Key) ([]kv.KeyRa
+	var taskStat rangetask.TaskStat
+	for {
+		select {
+		case <-ctx.Done():
+			return taskStat, errors.WithStack(ctx.Err())
+		default:
+		}
+
+		if len(rangeEndKey) > 0 && bytes.Compare(startKey, rangeEndKey) >= 0 {
+			break
+		}
+
+		bo := tikv.NewBackoffer(ctx, flashbackMaxBackoff)
+		loc, err := s.GetRegionCache().LocateKey(bo, startKey)
+		if err != nil {
+			return taskStat, err
+		}
+
+		endKey := loc.EndKey
+		isLast := len(endKey) == 0 || (len(rangeEndKey) > 0 && bytes.Compare(endKey, rangeEndKey) >= 0)
+		// If it is the last region
+		if isLast {
+			endKey = rangeEndKey
+		}
+
+		req := tikvrpc.NewRequest(tikvrpc.CmdFlashbackToVersion, &kvrpcpb.FlashbackToVersionRequest{
+			Version:  version,
+			StartKey: startKey,
+			EndKey:   endKey,
+		})
+
+		resp, err := s.SendReq(bo, req, loc.Region, flashbackTimeout)
+		if err != nil {
+			return taskStat, err
+		}
+		regionErr, err := resp.GetRegionError()
+		if err != nil {
+			return taskStat, err
+		}
+		if regionErr != nil {
+			err = bo.Backoff(tikv.BoRegionMiss(), errors.New(regionErr.String()))
+			if err != nil {
+				return taskStat, err
+			}
+			continue
+		}
+		if resp.Resp == nil {
+			return taskStat, errors.WithStack(tikverr.ErrBodyMissing)
+		}
+		flashbackToVersionResp := resp.Resp.(*kvrpcpb.FlashbackToVersionResponse)
+		if err := flashbackToVersionResp.GetError(); err != "" {
+			return taskStat, errors.Errorf("unexpected flashback to version err: %v", err)
+		}
+		taskStat.CompletedRegions++
+		if isLast {
+			break
+		}
+		startKey = endKey
+	}
+	return taskStat, nil
+}
+
+func flashbackToVersion(
+	ctx context.Context,
+	d *ddlCtx,
+	version uint64,
+	startKey []byte, endKey []byte,
+) (err error) {
+	return rangetask.NewRangeTaskRunner(
+		"flashback-to-version-runner",
+		d.store.(tikv.Storage),
+		int(variable.GetFlashbackConcurrency()),
+		func(ctx context.Context, r tikvstore.KeyRange) (rangetask.TaskStat, error) {
+			return sendFlashbackToVersionRPC(ctx, d.store.(tikv.Storage), version, r)
+		},
+	).RunOnRange(ctx, startKey, endKey)
+}
+
 // A Flashback has 3 different stages.
 // 1. before lock flashbackClusterJobID, check clusterJobID and lock it.
 // 2. before flashback start, check timestamp, disable GC and close PD schedule.
 // 3. before flashback done, get key ranges, send flashback RPC.
 func (w *worker) onFlashbackCluster(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, err error) {
+	inFlashbackTest := false
+	failpoint.Inject("mockFlashbackTest", func(val failpoint.Value) {
+		if val.(bool) {
+			inFlashbackTest = true
+		}
+	})
+	// TODO support flashback in unistore
+	if d.store.Name() != "TiKV" && !inFlashbackTest {
+		job.State = model.JobStateCancelled
+		return ver, errors.Errorf("Not support flashback cluster in non-TiKV env")
+	}
+
 	var flashbackTS uint64
 	var pdScheduleValue map[string]interface{}
-	if err := job.DecodeArgs(&flashbackTS, &pdScheduleValue); err != nil {
+	var readOnlyValue, gcEnabledValue bool
```

Comment:
Can we define `readOnlyValue` as `string`? So that we don't need a convert like:
```
	if value {
		setValue = variable.On
	} else {
		setValue = variable.Off
	}
```
In `setTiDBRestrictedReadOnly()`